src/sysinfo.cc

```diff
@@ -550,7 +550,9 @@ int GetNumCPUs() {
             "CPU ID assignments in /proc/cpuinfo seem messed up."
             " This is usually caused by a bad BIOS.\n");
   }
-  return num_cpus;
+  // If we are unable to identify the number of CPUs, default to one rather
+  // than zero.
```

Comment:
This shouldn't be a global fix-up.

Reply:
Can you elaborate what you mean?  It should continue to default to zero and there should be code to parse this nonstandard `/proc/cpuinfo`?  Or something else?

Reply:
> Presumably no other architecture so far has needed this fix-up, so if this fix-up is needed for alpha, it should not be retroactively added to other architectures at the same time.
> 
> If `/proc/cpuinfo` is different there, then you certainly should actually parse it in an alpha-specific way.

I believe there should be some sort of error condition then, rather than returning zero as if it was a valid number of CPUs.  The calling code does not seem to consider this an error condition, and just blindly keeps going.  For example, when returning zero CPUs, later code subtracts 1 from this (resulting in -1), then tries to cast to a `size_t` and allocate a vector of beyond `SIZE_MAX` length.  So, not being able to parse `/proc/cpuinfo` does not seem to have been considered as a valid error condition.

Would you be okay with having this throw a runtime error if `/proc/cpuinfo` is unable to be parsed?

Reply:
Once again, does that happen on Alpha, or on an other architecture?

Reply:
> Once again, does that happen on Alpha, or on an other architecture?

It happens on Alpha, but it would also happen on any architecture that uses a different format for `/proc/cpuinfo`.  For example, I just checked, and SPARC also uses a different format and runs into the same problem.

Yes, parsing for their formats should be added, but why not bail out when you don't have enough information rather than returning data you know is bad, especially if the callers don't check it?

Reply:
For example because it hides bugs. By the same logic,
`#error You need to define CycleTimer for your OS and CPU`
should not be there, but it should default to standard wall clock.
Either the architecture is supported, or it is not.

Reply:
> For example because it hides bugs. By the same logic, `#error You need to define CycleTimer for your OS and CPU` should not be there, but it should default to standard wall clock. Either the architecture is supported, or it is not.

Right, that's what I'm proposing.  Instead of defaulting to 0 CPUs or 1 CPU, throw an error instead (though it has to be a runtime error rather than compile-time).  Is that an OK solution?

Reply:
I suppose
```
  if (num_cpus == 0) {
    std::cerr << "Failed to detect CPU count\n";
    return -1;
  }
```
would make sense here, yes.
But not soft fix-up like it is now.

Reply:
> I suppose
> 
> ```
>   if (num_cpus == 0) {
>     std::cerr << "Failed to detect CPU count\n";
>     return -1;
>   }
> ```
> 
> would make sense here, yes. But not soft fix-up like it is now.

This won't work either, because callers don't check for errors.  I think it will have to be an exception.

Reply:
That is how errors are dealt with elsewhere in the function.

Reply:
we don't support exceptions.

Reply:
I just tested, and returning -1 results in the same issues as returning 0.  Here is an example of what it looks like with -1:

```
Running /var/tmp/portage/dev-cpp/benchmark-1.8.3/work/benchmark-1.8.3_build/test/benchmark_test
Run on (-1 X 1987.32 MHz CPU )
Load Average: 1.09, 1.04, 0.90
terminate called after throwing an instance of 'std::length_error'
  what():  cannot create std::vector larger than max_size()
<end of output>
Test time =   0.05 sec
----------------------------------------------------------
Test Failed.
```

I see that it's used immediately in `CPUInfo` constructor without error checking.  I'll have to experiment and see what is a reasonable way to propagate such an error up.

Reply:
FWIW, personally, i just wouldn't bother.

Reply:
Then how about `std::abort()` (or, I guess the abort handler stuff)?  It looks like there's just no good way for callers to handle errors here and there are no circumstances under which you can meaningfully work with zero CPUs.

Reply:
yeah we could.. the `CHECK` macros we have get compiled out in release builds but maybe we should have one that doesn't get compiled out (or just `CallAbortHandler` directly).